pkg/servicescm/servicescm.go

```diff
@@ -125,6 +134,150 @@ func Generate(name, namespace string) (*core.ConfigMap, error) {
+		return err
+	}
+	return validatePriorities(cmData.Services)
+}
+
+func validateRequiredContent(services []Service, files []FileInfo) error {
+	// Validate services
+	if len(services) != len(*requiredServices) {
+		return errors.New("Unexpected number of services")
+	}
+	for _, requiredSvc := range *requiredServices {
+		// isPresent checks if the required service exists within the given services slice
+		isPresent := func() error {
+			for _, svc := range services {
+				if reflect.DeepEqual(requiredSvc, svc) {
+					return nil
+				}
+			}
+			return errors.Errorf("Required service %s is not present with expected configuration", requiredSvc.Name)
+		}
+		if err := isPresent(); err != nil {
+			return err
+		}
+	}
+
+	// Validate files
+	if len(files) != len(*requiredFiles) {
+		return errors.New("Unexpected number of files")
+	}
+	for _, requiredFile := range *requiredFiles {
+		// isPresent checks if the required file exists within the given files slice
+		isPresent := func() error {
+			for _, file := range files {
+				if reflect.DeepEqual(requiredFile, file) {
+					return nil
+				}
+			}
+			return errors.Errorf("Required file %s is not present as expected", requiredFile.Path)
+		}
+		if err := isPresent(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// validateDependencies ensures that no bootstrap service depends on a non-bootstrap service
+func validateDependencies(services []Service) error {
+	bootstrapServices := []Service{}
+	nonBootstrapServices := []Service{}
+	for _, svc := range services {
+		if svc.Bootstrap {
+			bootstrapServices = append(bootstrapServices, svc)
+		} else {
+			nonBootstrapServices = append(nonBootstrapServices, svc)
+		}
+	}
+
+	for _, bootstrapSvc := range bootstrapServices {
+		if bootstrapSvc.hasDependency(nonBootstrapServices) {
+			return errors.Errorf("bootstrap service %s cannot depend on non-bootstrap service", bootstrapSvc.Name)
+		}
+	}
+	return nil
+}
+
+// hasDependency checks if a service is dependent on any services in the given slice
+func (s *Service) hasDependency(possibleDependencies []Service) bool {
+	for _, dependency := range s.Dependencies {
+		for _, possibleDependency := range possibleDependencies {
+			if dependency == possibleDependency.Name {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// validatePriorities ensures that each service that has the bootstrap flag set as true has a higher priority than all
+// non-bootstrap services. There should be no overlap in the priorities of bootstrap services and controller services.
+func validatePriorities(services []Service) error {
+	// sort services in ascending priority, bootstrap services towards the front of slice
+	sort.Slice(services, func(i, j int) bool {
+		return services[i].Priority < services[j].Priority
+	})
+
+	// ensure no bootstrap service appears after a controller service in the ordered list
+	nonBootstrapSeen := false
+	lastBootstrapPriority := uint(0)
+	for _, svc := range services {
```

Comment:
Moreover, if the sorted array is not used or needed elsewhere you could classify the `services` per `Priority` using a priority queue or hashmap and then validate, both will perform better than sorting the whole array (nLog(n) best case).

Reply:
I thought priority queue will serve the requirement better

Reply:
Had fun revisiting this, we could bring it down to O(n) with some more elaborate sorting (something like radix sort which is O(n) then the O(n) traversal to find the first nonBootstrap service). 
AFAIK there is no standard implementation of priority queue in go and I don't see our use case calling for a custom priority queue implementation, so I will leave it as is.

Reply:
I agree. we need to implement a few functions. I am good with nlogn also. 
BTW I am assuming a max of 5 to 6 services now and expecting it might reach 10-15 to the max. So sorting shouldn't impact performance that much.

Reply:
Glad everyone is thinking about complexity. Given we are never going to hit 1000s of services, I am happy that you are keeping this simple.